### ISSUE TITLE
Sandbox: ENOPRO: No file system provider found for resource 'vscode-userdata:/sandbox-user-data-dir/window.log'

### ISSUE THREAD

**Issue by bpasero** [microsoft/vscode#116829]
Steps to Reproduce:

1. run out of sources `scripts/code.sh --__sandbox`
2. see the cli output from the terminal where you launched

```
[main 2021-02-17T05:50:19.205Z] [uncaught exception in main]: ENOPRO: No file system provider found for resource 'vscode-userdata:/sandbox-user-data-dir/window.log'
[main 2021-02-17T05:50:19.205Z] ENOPRO: No file system provider found for resource 'vscode-userdata:/sandbox-user-data-dir/window.log'
    at FileService.withProvider (/Users/bpasero/Development/Microsoft/vscode-electron-12.x/out/vs/platform/files/common/fileService.js:105:31)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at async FileService.exists (/Users/bpasero/Development/Microsoft/vscode-electron-12.x/out/vs/platform/files/common/fileService.js:219:30)
    at async FileService.doValidateCreateFile (/Users/bpasero/Development/Microsoft/vscode-electron-12.x/out/vs/platform/files/common/fileService.js:241:91)
    at async FileService.createFile (/Users/bpasero/Development/Microsoft/vscode-electron-12.x/out/vs/platform/files/common/fileService.js:247:13)
    at async FileLogger.initialize (/Users/bpasero/Development/Microsoft/vscode-electron-12.x/out/vs/platform/log/common/fileLog.js:76:17)
```

It looks like `electron-main` tries to use the `vscode-userdata` protocol which is likely not registered there?

**Comment by sandy081** [comment]
@bpasero It seems you have enabled logging in sandbox - f49d78c4afc16c4725f439be610986c9a8202b1b which caused this. We can enable real logging only when file protocol is supported. So I reverted that to use simple logging.

**Comment by bpasero** [comment]
@sandy081 sorry, I do not follow the reasoning. It is true that we have no file system provider for local file access in sandbox, but for that reason I register a in memory fs provider here:

https://github.com/microsoft/vscode/blob/29d4fdb65c8bae2001d1dc3178737127fde3d14d/src/vs/workbench/electron-sandbox/desktop.main.ts#L196-L196

The exception actually seems to happen in the main process, so there must be code that is not triggered when sandbox is enabled? Can you explain this further?

**Comment by sandy081** [comment]
@bpasero Exception you saw in the main process was because the environment service is pointing to a log file with userData scheme

https://github.com/microsoft/vscode/blob/29d4fdb65c8bae2001d1dc3178737127fde3d14d/src/vs/workbench/electron-sandbox/sandbox.simpleservices.ts#L74-L74

Since logging happens in the main now, window asks the main for the logger with this URI. Main does not have fsp for userData scheme, so it is throwing the above error.

I tired changing the window log file in sandbox workbench environment service to use the real log path (with file scheme). It seems that the simple fsp is not supporting this and logs are not shown in output channel because output channel use file system service to load and show logs. Hence I did not push that. 

In order to have logging completely work in sandbox, it needs the real fsp that supports file protocol in the window. 

Hope it is clear.

**Comment by bpasero** [comment]
@sandy081 in sandbox we register a simple file system provider:

https://github.com/microsoft/vscode/blob/29d4fdb65c8bae2001d1dc3178737127fde3d14d/src/vs/workbench/electron-sandbox/desktop.main.ts#L196-L196

Could it be that you have some native node.js fs dependencies still that we haven't looked at?

**Comment by sandy081** [comment]
@bpasero 

To show log file in the output channel, I wait until the corresponding log file is created and after that I register the output log channel for the log file.

In sandbox simple workbench environment service, If I change the window log file to use the real log file on disk then all window logging happens on main ie., spdlog in main writes into the log file. If I understand correctly in sandbox window, simple fsp is based on `InMemoryFileSystemProvider` and there is no FSP that supports files on disk as of now. So even though the logging works correctly but output channel cannot show the log file because of no FSP for files on disk.

Hence I thought, I would enable real logging in sandbox when we add FSP support for files on disk.

**Comment by bpasero** [comment]
@sandy081 nevermind, I figured out the root cause: we pick up on a `IEnvironmentService#logFile` variable and in sandbox it is defined as:

https://github.com/microsoft/vscode/blob/29d4fdb65c8bae2001d1dc3178737127fde3d14d/src/vs/workbench/electron-sandbox/sandbox.simpleservices.ts#L74-L74

However, since now the main side is responsible for writing the logs, main will try to write to a `userData` scheme because that is how `userRoamingDataHome` is defined in sandbox. 

Bottom line, I think the path for logs need to be `file` even in sandbox land because they travel to the main side. I wonder if in that case these environment properties should rather be `string` and not `URI` because we seem to only accept file paths anyways?

**Comment by sandy081** [comment]
> Bottom line, I think the path for logs need to be file even in sandbox land because they travel to the main side. I wonder if in that case these environment properties should rather be string and not URI because we seem to only accept file paths anyways?

Its not entirely true. In web they are not file scheme instead `vscode-log` scheme. Hence log paths are URIs. They are also used at multiple places in workbench 

- output 

https://github.com/microsoft/vscode/blob/47a956eb2e074720cdc51aef74ba61d628061c54/src/vs/workbench/contrib/logs/common/logs.contribution.ts#L49-L49

- extension host

**Comment by bpasero** [comment]
Ah yeah fair enough.
